src/interactivity/win32/ut_interactivity_win32/UiaTextRangeTests.cpp

```diff
@@ -19,13 +21,193 @@ using namespace Microsoft::WRL;
+                break;
+            }
+            else if (pos_x == bounds.left())
+            {
+                // left boundary --> wrap
+                pos_x = bounds.right() - 1;
+                --pos_y;
+            }
+            else
+            {
+                // standard move
+                --pos_x;
+            }
+            ++amt;
+        }
+    }
+    return { pos_x, pos_y };
+}
+
+static constexpr til::point point_offset_by_line(const til::point start, const til::rectangle bounds, ptrdiff_t amt)
+{
+    // X = left boundary for UIA
+    ptrdiff_t pos_x = bounds.left();
+    ptrdiff_t pos_y = start.y();
+    while (amt != 0)
+    {
+        if (amt > 0)
+        {
+            if (pos_y == bounds.bottom() + 1)
+            {
+                break;
+            }
+            else
+            {
+                ++pos_y;
+            }
+            --amt;
+        }
+        else
+        {
+            if (pos_y == bounds.top())
+            {
+                break;
+            }
+            else
+            {
+                --pos_y;
+            }
+            ++amt;
+        }
+    }
+    return { pos_x, pos_y };
+}
+
+// IMPORTANT: reference this _after_ defining point_offset_by_XXX. We need it for some definitions
+#include "GeneratedUiaTextRangeMovementTests.g.cpp"
+
+namespace
+{
+#pragma region TAEF hookup for the test case array above
+    struct ArrayIndexTaefAdapterRow : public Microsoft::WRL::RuntimeClass<Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::ClassicCom | Microsoft::WRL::InhibitFtmBase>, IDataRow>
+    {
+        HRESULT RuntimeClassInitialize(const size_t index)
+        {
+            _index = index;
+            return S_OK;
+        }
+
+        STDMETHODIMP GetTestData(BSTR /*pszName*/, SAFEARRAY** ppData) override
+        {
+            const auto indexString{ wil::str_printf<std::wstring>(L"%zu", _index) };
+            auto safeArray{ SafeArrayCreateVector(VT_BSTR, 0, 1) };
+            LONG index{ 0 };
+            auto indexBstr{ wil::make_bstr(indexString.c_str()) };
+            (void)SafeArrayPutElement(safeArray, &index, indexBstr.release());
+            *ppData = safeArray;
+            return S_OK;
+        }
+
+        STDMETHODIMP GetMetadataNames(SAFEARRAY** ppMetadataNames) override
+        {
+            *ppMetadataNames = nullptr;
+            return S_FALSE;
+        }
+
+        STDMETHODIMP GetMetadata(BSTR /*pszName*/, SAFEARRAY** ppData) override
+        {
+            *ppData = nullptr;
+            return S_FALSE;
+        }
+
+        STDMETHODIMP GetName(BSTR* ppszRowName) override
+        {
+            *ppszRowName = wil::make_bstr(s_movementTests[_index].name.data()).release();
+            return S_OK;
+        }
+
+    private:
+        size_t _index;
+    };
+
+    struct ArrayIndexTaefAdapterSource : public Microsoft::WRL::RuntimeClass<Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::ClassicCom | Microsoft::WRL::InhibitFtmBase>, IDataSource>

```

Comment:
Would it be reasonable to use `<winrt/base.h>` instead?

Reply:
Maybe? I stole this from https://github.com/microsoft/terminal/blob/main/src/buffer/out/ut_textbuffer/ReflowTests.cpp.

I'll leave this as-is for now since we'll want to consolidate these at some point.